// @ant-design/pro-layout/lib/components/Header/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard").default;
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultHeader = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _proProvider = require("@ant-design/pro-provider");
var _antd = require("antd");
var _classnames = _interopRequireDefault(require("classnames"));
var _react = _interopRequireWildcard(require("react"));
var _header = require("./style/header");
var _utils = require("../../utils/utils");
var _GlobalHeader = require("../GlobalHeader");
var _TopNavHeader = require("../TopNavHeader");
var _stylish = require("./style/stylish");
var _jsxRuntime = require("react/jsx-runtime");
var Header = _antd.Layout.Header;
var DefaultHeader = function DefaultHeader(props) {
  var _classNames, _token$layout, _token$layout$header, _token$layout2, _token$layout2$header;
  var isMobile = props.isMobile,
    fixedHeader = props.fixedHeader,
    propsClassName = props.className,
    style = props.style,
    collapsed = props.collapsed,
    prefixCls = props.prefixCls,
    onCollapse = props.onCollapse,
    layout = props.layout,
    headerRender = props.headerRender,
    headerContentRender = props.headerContentRender;
  var _useContext = (0, _react.useContext)(_proProvider.ProProvider),
    token = _useContext.token;
  var renderContent = (0, _react.useCallback)(function () {
    var isTop = layout === 'top';
    var clearMenuData = (0, _utils.clearMenuItem)(props.menuData || []);
    var defaultDom = /*#__PURE__*/(0, _jsxRuntime.jsx)(_GlobalHeader.GlobalHeader, (0, _objectSpread2.default)((0, _objectSpread2.default)({
      onCollapse: onCollapse
    }, props), {}, {
      menuData: clearMenuData,
      children: headerContentRender && headerContentRender(props, null)
    }));
    if (isTop && !isMobile) {
      defaultDom = /*#__PURE__*/(0, _jsxRuntime.jsx)(_TopNavHeader.TopNavHeader, (0, _objectSpread2.default)((0, _objectSpread2.default)({
        mode: "horizontal",
        onCollapse: onCollapse
      }, props), {}, {
        menuData: clearMenuData
      }));
    }
    if (headerRender && typeof headerRender === 'function') {
      return headerRender(props, defaultDom);
    }
    return defaultDom;
  }, [headerContentRender, headerRender, isMobile, layout, onCollapse, props]);
  var needFixedHeader = fixedHeader || layout === 'mix';
  var isTop = layout === 'top';
  var baseClassName = "".concat(prefixCls, "-layout-header");
  var _useStyle = (0, _header.useStyle)(baseClassName),
    wrapSSR = _useStyle.wrapSSR,
    hashId = _useStyle.hashId;
  var stylish = (0, _stylish.useStylish)("".concat(baseClassName, ".").concat(baseClassName, "-stylish"), {
    proLayoutCollapsedWidth: 64,
    stylish: props.stylish
  });
  var className = (0, _classnames.default)(propsClassName, hashId, baseClassName, (_classNames = {}, (0, _defineProperty2.default)(_classNames, "".concat(baseClassName, "-fixed-header"), needFixedHeader), (0, _defineProperty2.default)(_classNames, "".concat(baseClassName, "-mix"), layout === 'mix'), (0, _defineProperty2.default)(_classNames, "".concat(baseClassName, "-fixed-header-action"), !collapsed), (0, _defineProperty2.default)(_classNames, "".concat(baseClassName, "-top-menu"), isTop), (0, _defineProperty2.default)(_classNames, "".concat(baseClassName, "-header"), true), (0, _defineProperty2.default)(_classNames, "".concat(baseClassName, "-stylish"), !!props.stylish), _classNames));
  if (layout === 'side' && !isMobile) return null;
  return stylish.wrapSSR(wrapSSR( /*#__PURE__*/(0, _jsxRuntime.jsx)(_jsxRuntime.Fragment, {
    children: /*#__PURE__*/(0, _jsxRuntime.jsxs)(_antd.ConfigProvider
    // @ts-ignore
    , {
      theme: {
        hashed: (0, _proProvider.isNeedOpenHash)(),
        components: {
          Layout: {
            colorBgHeader: 'transparent',
            colorBgBody: 'transparent'
          }
        }
      },
      children: [needFixedHeader && /*#__PURE__*/(0, _jsxRuntime.jsx)(Header, {
        style: (0, _objectSpread2.default)({
          height: (token === null || token === void 0 ? void 0 : (_token$layout = token.layout) === null || _token$layout === void 0 ? void 0 : (_token$layout$header = _token$layout.header) === null || _token$layout$header === void 0 ? void 0 : _token$layout$header.heightLayoutHeader) || 56,
          lineHeight: "".concat((token === null || token === void 0 ? void 0 : (_token$layout2 = token.layout) === null || _token$layout2 === void 0 ? void 0 : (_token$layout2$header = _token$layout2.header) === null || _token$layout2$header === void 0 ? void 0 : _token$layout2$header.heightLayoutHeader) || 56, "px"),
          backgroundColor: 'transparent',
          zIndex: 19
        }, style)
      }), /*#__PURE__*/(0, _jsxRuntime.jsx)(Header, {
        className: className,
        style: style,
        children: renderContent()
      })]
    })
  })));
};
exports.DefaultHeader = DefaultHeader;
