// @ant-design/pro-layout/lib/components/SiderMenu/style/menu.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useStyle = useStyle;
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _proProvider = require("@ant-design/pro-provider");
var genProLayoutBaseMenuStyle = function genProLayoutBaseMenuStyle(token, mode) {
  var _token$layout, _token$layout2, _collapsed, _collapsed2, _$concat3, _itemTitle, _$concat5;
  var menuToken = mode.includes('horizontal') ? (_token$layout = token.layout) === null || _token$layout === void 0 ? void 0 : _token$layout.header : (_token$layout2 = token.layout) === null || _token$layout2 === void 0 ? void 0 : _token$layout2.sider;
  return (0, _defineProperty2.default)({}, "".concat(token.componentCls), (_$concat5 = {
    background: 'transparent',
    color: menuToken === null || menuToken === void 0 ? void 0 : menuToken.colorTextMenu,
    border: 'none'
  }, (0, _defineProperty2.default)(_$concat5, "".concat(token.componentCls, "-menu-item"), {
    transition: 'none !important'
  }), (0, _defineProperty2.default)(_$concat5, "".concat(token.componentCls, "-submenu-has-icon"), (0, _defineProperty2.default)({}, "> ".concat(token.antCls, "-menu-sub"), {
    paddingInlineStart: 10
  })), (0, _defineProperty2.default)(_$concat5, "".concat(token.antCls, "-menu-title-content"), {
    width: '100%',
    height: '100%',
    display: 'inline-flex'
  }), (0, _defineProperty2.default)(_$concat5, "".concat(token.antCls, "-menu-title-content"), {
    '&:first-child': {
      width: '100%'
    }
  }), (0, _defineProperty2.default)(_$concat5, "&&-collapsed", (_collapsed = {}, (0, _defineProperty2.default)(_collapsed, "".concat(token.antCls, "-menu-item, \n        ").concat(token.antCls, "-menu-item-group > ").concat(token.antCls, "-menu-item-group-list > ").concat(token.antCls, "-menu-item, \n        ").concat(token.antCls, "-menu-item-group > ").concat(token.antCls, "-menu-item-group-list > ").concat(token.antCls, "-menu-submenu > ").concat(token.antCls, "-menu-submenu-title, \n        ").concat(token.antCls, "-menu-submenu > ").concat(token.antCls, "-menu-submenu-title"), {
    paddingInline: '0 !important',
    marginBlock: '4px !important'
  }), (0, _defineProperty2.default)(_collapsed, "".concat(token.antCls, "-menu-item-group > ").concat(token.antCls, "-menu-item-group-list > ").concat(token.antCls, "-menu-submenu-selected > ").concat(token.antCls, "-menu-submenu-title, \n        ").concat(token.antCls, "-menu-submenu-selected > ").concat(token.antCls, "-menu-submenu-title"), {
    backgroundColor: menuToken === null || menuToken === void 0 ? void 0 : menuToken.colorBgMenuItemSelected,
    borderRadius: token.borderRadiusLG
  }), (0, _defineProperty2.default)(_collapsed, "".concat(token.componentCls, "-group"), (0, _defineProperty2.default)({}, "".concat(token.antCls, "-menu-item-group-title"), {
    paddingInline: 0
  })), _collapsed)), (0, _defineProperty2.default)(_$concat5, "".concat(token.componentCls, "-item-icon"), {
    height: '14px',
    width: '14px',
    opacity: '0.85',
    lineHeight: '14px',
    '> span.anticon': {
      lineHeight: '14px!important',
      height: '14px'
    }
  }), (0, _defineProperty2.default)(_$concat5, '& &-item-title', (_itemTitle = {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    gap: 8,
    color: menuToken === null || menuToken === void 0 ? void 0 : menuToken.colorTextMenu,
    '&-collapsed': (_collapsed2 = {
      flexDirection: 'column',
      justifyContent: 'center',
      minWidth: 40,
      height: 40
    }, (0, _defineProperty2.default)(_collapsed2, "".concat(token.componentCls, "-item-icon"), {
      height: '16px',
      width: '16px',
      lineHeight: '16px !important',
      '.anticon': {
        lineHeight: '16px',
        height: '16px'
      }
    }), (0, _defineProperty2.default)(_collapsed2, "".concat(token.componentCls, "-item-text"), {
      maxWidth: '100%'
    }), (0, _defineProperty2.default)(_collapsed2, "".concat(token.componentCls, "-item-text-has-icon"), {
      display: 'none !important'
    }), _collapsed2)
  }, (0, _defineProperty2.default)(_itemTitle, "&".concat(token.componentCls, "-group-item-title"), {
    gap: 4,
    height: 18,
    overflow: 'hidden'
  }), (0, _defineProperty2.default)(_itemTitle, "&".concat(token.componentCls, "-item-collapsed-show-title"), (0, _defineProperty2.default)({
    lineHeight: '16px',
    height: '48px'
  }, "&".concat(token.componentCls, "-item-title-collapsed"), (_$concat3 = {
    display: 'flex'
  }, (0, _defineProperty2.default)(_$concat3, "".concat(token.componentCls, "-item-icon"), {
    height: '16px',
    width: '16px',
    lineHeight: '16px !important',
    '.anticon': {
      lineHeight: '16px',
      height: '16px'
    }
  }), (0, _defineProperty2.default)(_$concat3, "".concat(token.componentCls, "-item-text"), {
    opacity: '1 !important',
    display: 'inline !important',
    textAlign: 'center',
    fontSize: 12,
    height: 12,
    lineHeight: '12px',
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
    width: '100%',
    margin: 0,
    padding: 0,
    marginBlockStart: 4
  }), _$concat3))), _itemTitle)), (0, _defineProperty2.default)(_$concat5, '&-group', (0, _defineProperty2.default)({}, "".concat(token.antCls, "-menu-item-group-title"), {
    fontSize: 12,
    color: token.colorTextLabel,
    '.anticon': {
      marginInlineEnd: 8
    }
  })), (0, _defineProperty2.default)(_$concat5, '&-group-divider', {
    color: token.colorTextSecondary,
    fontSize: 12,
    lineHeight: 20
  }), _$concat5));
};
function useStyle(prefixCls, mode) {
  return (0, _proProvider.useStyle)('ProLayoutBaseMenu' + mode, function (token) {
    var proLayoutMenuToken = (0, _objectSpread2.default)((0, _objectSpread2.default)({}, token), {}, {
      componentCls: ".".concat(prefixCls)
    });
    return [genProLayoutBaseMenuStyle(proLayoutMenuToken, mode || 'inline')];
  });
}
