// ahooks/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "clearCache", {
  enumerable: true,
  get: function get() {
    return _useRequest.clearCache;
  }
});
Object.defineProperty(exports, "configResponsive", {
  enumerable: true,
  get: function get() {
    return _useResponsive.configResponsive;
  }
});
Object.defineProperty(exports, "createUpdateEffect", {
  enumerable: true,
  get: function get() {
    return _createUpdateEffect.createUpdateEffect;
  }
});
Object.defineProperty(exports, "useAntdTable", {
  enumerable: true,
  get: function get() {
    return _useAntdTable["default"];
  }
});
Object.defineProperty(exports, "useAsyncEffect", {
  enumerable: true,
  get: function get() {
    return _useAsyncEffect["default"];
  }
});
Object.defineProperty(exports, "useBoolean", {
  enumerable: true,
  get: function get() {
    return _useBoolean["default"];
  }
});
Object.defineProperty(exports, "useClickAway", {
  enumerable: true,
  get: function get() {
    return _useClickAway["default"];
  }
});
Object.defineProperty(exports, "useControllableValue", {
  enumerable: true,
  get: function get() {
    return _useControllableValue["default"];
  }
});
Object.defineProperty(exports, "useCookieState", {
  enumerable: true,
  get: function get() {
    return _useCookieState["default"];
  }
});
Object.defineProperty(exports, "useCountDown", {
  enumerable: true,
  get: function get() {
    return _useCountDown["default"];
  }
});
Object.defineProperty(exports, "useCounter", {
  enumerable: true,
  get: function get() {
    return _useCounter["default"];
  }
});
Object.defineProperty(exports, "useCreation", {
  enumerable: true,
  get: function get() {
    return _useCreation["default"];
  }
});
Object.defineProperty(exports, "useDebounce", {
  enumerable: true,
  get: function get() {
    return _useDebounce["default"];
  }
});
Object.defineProperty(exports, "useDebounceEffect", {
  enumerable: true,
  get: function get() {
    return _useDebounceEffect["default"];
  }
});
Object.defineProperty(exports, "useDebounceFn", {
  enumerable: true,
  get: function get() {
    return _useDebounceFn["default"];
  }
});
Object.defineProperty(exports, "useDeepCompareEffect", {
  enumerable: true,
  get: function get() {
    return _useDeepCompareEffect["default"];
  }
});
Object.defineProperty(exports, "useDeepCompareLayoutEffect", {
  enumerable: true,
  get: function get() {
    return _useDeepCompareLayoutEffect["default"];
  }
});
Object.defineProperty(exports, "useDocumentVisibility", {
  enumerable: true,
  get: function get() {
    return _useDocumentVisibility["default"];
  }
});
Object.defineProperty(exports, "useDrag", {
  enumerable: true,
  get: function get() {
    return _useDrag["default"];
  }
});
Object.defineProperty(exports, "useDrop", {
  enumerable: true,
  get: function get() {
    return _useDrop["default"];
  }
});
Object.defineProperty(exports, "useDynamicList", {
  enumerable: true,
  get: function get() {
    return _useDynamicList["default"];
  }
});
Object.defineProperty(exports, "useEventEmitter", {
  enumerable: true,
  get: function get() {
    return _useEventEmitter["default"];
  }
});
Object.defineProperty(exports, "useEventListener", {
  enumerable: true,
  get: function get() {
    return _useEventListener["default"];
  }
});
Object.defineProperty(exports, "useEventTarget", {
  enumerable: true,
  get: function get() {
    return _useEventTarget["default"];
  }
});
Object.defineProperty(exports, "useExternal", {
  enumerable: true,
  get: function get() {
    return _useExternal["default"];
  }
});
Object.defineProperty(exports, "useFavicon", {
  enumerable: true,
  get: function get() {
    return _useFavicon["default"];
  }
});
Object.defineProperty(exports, "useFocusWithin", {
  enumerable: true,
  get: function get() {
    return _useFocusWithin["default"];
  }
});
Object.defineProperty(exports, "useFullscreen", {
  enumerable: true,
  get: function get() {
    return _useFullscreen["default"];
  }
});
Object.defineProperty(exports, "useFusionTable", {
  enumerable: true,
  get: function get() {
    return _useFusionTable["default"];
  }
});
Object.defineProperty(exports, "useGetState", {
  enumerable: true,
  get: function get() {
    return _useGetState["default"];
  }
});
Object.defineProperty(exports, "useHistoryTravel", {
  enumerable: true,
  get: function get() {
    return _useHistoryTravel["default"];
  }
});
Object.defineProperty(exports, "useHover", {
  enumerable: true,
  get: function get() {
    return _useHover["default"];
  }
});
Object.defineProperty(exports, "useInViewport", {
  enumerable: true,
  get: function get() {
    return _useInViewport["default"];
  }
});
Object.defineProperty(exports, "useInfiniteScroll", {
  enumerable: true,
  get: function get() {
    return _useInfiniteScroll["default"];
  }
});
Object.defineProperty(exports, "useInterval", {
  enumerable: true,
  get: function get() {
    return _useInterval["default"];
  }
});
Object.defineProperty(exports, "useIsomorphicLayoutEffect", {
  enumerable: true,
  get: function get() {
    return _useIsomorphicLayoutEffect["default"];
  }
});
Object.defineProperty(exports, "useKeyPress", {
  enumerable: true,
  get: function get() {
    return _useKeyPress["default"];
  }
});
Object.defineProperty(exports, "useLatest", {
  enumerable: true,
  get: function get() {
    return _useLatest["default"];
  }
});
Object.defineProperty(exports, "useLocalStorageState", {
  enumerable: true,
  get: function get() {
    return _useLocalStorageState["default"];
  }
});
Object.defineProperty(exports, "useLockFn", {
  enumerable: true,
  get: function get() {
    return _useLockFn["default"];
  }
});
Object.defineProperty(exports, "useLongPress", {
  enumerable: true,
  get: function get() {
    return _useLongPress["default"];
  }
});
Object.defineProperty(exports, "useMap", {
  enumerable: true,
  get: function get() {
    return _useMap["default"];
  }
});
Object.defineProperty(exports, "useMemoizedFn", {
  enumerable: true,
  get: function get() {
    return _useMemoizedFn["default"];
  }
});
Object.defineProperty(exports, "useMount", {
  enumerable: true,
  get: function get() {
    return _useMount["default"];
  }
});
Object.defineProperty(exports, "useMouse", {
  enumerable: true,
  get: function get() {
    return _useMouse["default"];
  }
});
Object.defineProperty(exports, "useMutationObserver", {
  enumerable: true,
  get: function get() {
    return _useMutationObserver["default"];
  }
});
Object.defineProperty(exports, "useNetwork", {
  enumerable: true,
  get: function get() {
    return _useNetwork["default"];
  }
});
Object.defineProperty(exports, "usePagination", {
  enumerable: true,
  get: function get() {
    return _usePagination["default"];
  }
});
Object.defineProperty(exports, "usePrevious", {
  enumerable: true,
  get: function get() {
    return _usePrevious["default"];
  }
});
Object.defineProperty(exports, "useRafInterval", {
  enumerable: true,
  get: function get() {
    return _useRafInterval["default"];
  }
});
Object.defineProperty(exports, "useRafState", {
  enumerable: true,
  get: function get() {
    return _useRafState["default"];
  }
});
Object.defineProperty(exports, "useRafTimeout", {
  enumerable: true,
  get: function get() {
    return _useRafTimeout["default"];
  }
});
Object.defineProperty(exports, "useReactive", {
  enumerable: true,
  get: function get() {
    return _useReactive["default"];
  }
});
Object.defineProperty(exports, "useRequest", {
  enumerable: true,
  get: function get() {
    return _useRequest["default"];
  }
});
Object.defineProperty(exports, "useResetState", {
  enumerable: true,
  get: function get() {
    return _useResetState["default"];
  }
});
Object.defineProperty(exports, "useResponsive", {
  enumerable: true,
  get: function get() {
    return _useResponsive.useResponsive;
  }
});
Object.defineProperty(exports, "useSafeState", {
  enumerable: true,
  get: function get() {
    return _useSafeState["default"];
  }
});
Object.defineProperty(exports, "useScroll", {
  enumerable: true,
  get: function get() {
    return _useScroll["default"];
  }
});
Object.defineProperty(exports, "useSelections", {
  enumerable: true,
  get: function get() {
    return _useSelections["default"];
  }
});
Object.defineProperty(exports, "useSessionStorageState", {
  enumerable: true,
  get: function get() {
    return _useSessionStorageState["default"];
  }
});
Object.defineProperty(exports, "useSet", {
  enumerable: true,
  get: function get() {
    return _useSet["default"];
  }
});
Object.defineProperty(exports, "useSetState", {
  enumerable: true,
  get: function get() {
    return _useSetState["default"];
  }
});
Object.defineProperty(exports, "useSize", {
  enumerable: true,
  get: function get() {
    return _useSize["default"];
  }
});
Object.defineProperty(exports, "useTextSelection", {
  enumerable: true,
  get: function get() {
    return _useTextSelection["default"];
  }
});
Object.defineProperty(exports, "useThrottle", {
  enumerable: true,
  get: function get() {
    return _useThrottle["default"];
  }
});
Object.defineProperty(exports, "useThrottleEffect", {
  enumerable: true,
  get: function get() {
    return _useThrottleEffect["default"];
  }
});
Object.defineProperty(exports, "useThrottleFn", {
  enumerable: true,
  get: function get() {
    return _useThrottleFn["default"];
  }
});
Object.defineProperty(exports, "useTimeout", {
  enumerable: true,
  get: function get() {
    return _useTimeout["default"];
  }
});
Object.defineProperty(exports, "useTitle", {
  enumerable: true,
  get: function get() {
    return _useTitle["default"];
  }
});
Object.defineProperty(exports, "useToggle", {
  enumerable: true,
  get: function get() {
    return _useToggle["default"];
  }
});
Object.defineProperty(exports, "useTrackedEffect", {
  enumerable: true,
  get: function get() {
    return _useTrackedEffect["default"];
  }
});
Object.defineProperty(exports, "useUnmount", {
  enumerable: true,
  get: function get() {
    return _useUnmount["default"];
  }
});
Object.defineProperty(exports, "useUnmountedRef", {
  enumerable: true,
  get: function get() {
    return _useUnmountedRef["default"];
  }
});
Object.defineProperty(exports, "useUpdate", {
  enumerable: true,
  get: function get() {
    return _useUpdate["default"];
  }
});
Object.defineProperty(exports, "useUpdateEffect", {
  enumerable: true,
  get: function get() {
    return _useUpdateEffect["default"];
  }
});
Object.defineProperty(exports, "useUpdateLayoutEffect", {
  enumerable: true,
  get: function get() {
    return _useUpdateLayoutEffect["default"];
  }
});
Object.defineProperty(exports, "useVirtualList", {
  enumerable: true,
  get: function get() {
    return _useVirtualList["default"];
  }
});
Object.defineProperty(exports, "useWebSocket", {
  enumerable: true,
  get: function get() {
    return _useWebSocket["default"];
  }
});
Object.defineProperty(exports, "useWhyDidYouUpdate", {
  enumerable: true,
  get: function get() {
    return _useWhyDidYouUpdate["default"];
  }
});
var _createUpdateEffect = require("./createUpdateEffect");
var _useAntdTable = _interopRequireDefault(require("./useAntdTable"));
var _useAsyncEffect = _interopRequireDefault(require("./useAsyncEffect"));
var _useBoolean = _interopRequireDefault(require("./useBoolean"));
var _useClickAway = _interopRequireDefault(require("./useClickAway"));
var _useControllableValue = _interopRequireDefault(require("./useControllableValue"));
var _useCookieState = _interopRequireDefault(require("./useCookieState"));
var _useCountDown = _interopRequireDefault(require("./useCountDown"));
var _useCounter = _interopRequireDefault(require("./useCounter"));
var _useCreation = _interopRequireDefault(require("./useCreation"));
var _useDebounce = _interopRequireDefault(require("./useDebounce"));
var _useDebounceEffect = _interopRequireDefault(require("./useDebounceEffect"));
var _useDebounceFn = _interopRequireDefault(require("./useDebounceFn"));
var _useDeepCompareEffect = _interopRequireDefault(require("./useDeepCompareEffect"));
var _useDeepCompareLayoutEffect = _interopRequireDefault(require("./useDeepCompareLayoutEffect"));
var _useDocumentVisibility = _interopRequireDefault(require("./useDocumentVisibility"));
var _useDrag = _interopRequireDefault(require("./useDrag"));
var _useDrop = _interopRequireDefault(require("./useDrop"));
var _useDynamicList = _interopRequireDefault(require("./useDynamicList"));
var _useEventEmitter = _interopRequireDefault(require("./useEventEmitter"));
var _useEventListener = _interopRequireDefault(require("./useEventListener"));
var _useEventTarget = _interopRequireDefault(require("./useEventTarget"));
var _useExternal = _interopRequireDefault(require("./useExternal"));
var _useFavicon = _interopRequireDefault(require("./useFavicon"));
var _useFocusWithin = _interopRequireDefault(require("./useFocusWithin"));
var _useFullscreen = _interopRequireDefault(require("./useFullscreen"));
var _useFusionTable = _interopRequireDefault(require("./useFusionTable"));
var _useGetState = _interopRequireDefault(require("./useGetState"));
var _useHistoryTravel = _interopRequireDefault(require("./useHistoryTravel"));
var _useHover = _interopRequireDefault(require("./useHover"));
var _useInfiniteScroll = _interopRequireDefault(require("./useInfiniteScroll"));
var _useInterval = _interopRequireDefault(require("./useInterval"));
var _useInViewport = _interopRequireDefault(require("./useInViewport"));
var _useIsomorphicLayoutEffect = _interopRequireDefault(require("./useIsomorphicLayoutEffect"));
var _useKeyPress = _interopRequireDefault(require("./useKeyPress"));
var _useLatest = _interopRequireDefault(require("./useLatest"));
var _useLocalStorageState = _interopRequireDefault(require("./useLocalStorageState"));
var _useLockFn = _interopRequireDefault(require("./useLockFn"));
var _useLongPress = _interopRequireDefault(require("./useLongPress"));
var _useMap = _interopRequireDefault(require("./useMap"));
var _useMemoizedFn = _interopRequireDefault(require("./useMemoizedFn"));
var _useMount = _interopRequireDefault(require("./useMount"));
var _useMouse = _interopRequireDefault(require("./useMouse"));
var _useNetwork = _interopRequireDefault(require("./useNetwork"));
var _usePagination = _interopRequireDefault(require("./usePagination"));
var _usePrevious = _interopRequireDefault(require("./usePrevious"));
var _useRafInterval = _interopRequireDefault(require("./useRafInterval"));
var _useRafState = _interopRequireDefault(require("./useRafState"));
var _useRafTimeout = _interopRequireDefault(require("./useRafTimeout"));
var _useReactive = _interopRequireDefault(require("./useReactive"));
var _useRequest = _interopRequireWildcard(require("./useRequest"));
var _useResetState = _interopRequireDefault(require("./useResetState"));
var _useResponsive = require("./useResponsive");
var _useSafeState = _interopRequireDefault(require("./useSafeState"));
var _useScroll = _interopRequireDefault(require("./useScroll"));
var _useSelections = _interopRequireDefault(require("./useSelections"));
var _useSessionStorageState = _interopRequireDefault(require("./useSessionStorageState"));
var _useSet = _interopRequireDefault(require("./useSet"));
var _useSetState = _interopRequireDefault(require("./useSetState"));
var _useSize = _interopRequireDefault(require("./useSize"));
var _useTextSelection = _interopRequireDefault(require("./useTextSelection"));
var _useThrottle = _interopRequireDefault(require("./useThrottle"));
var _useThrottleEffect = _interopRequireDefault(require("./useThrottleEffect"));
var _useThrottleFn = _interopRequireDefault(require("./useThrottleFn"));
var _useTimeout = _interopRequireDefault(require("./useTimeout"));
var _useTitle = _interopRequireDefault(require("./useTitle"));
var _useToggle = _interopRequireDefault(require("./useToggle"));
var _useTrackedEffect = _interopRequireDefault(require("./useTrackedEffect"));
var _useUnmount = _interopRequireDefault(require("./useUnmount"));
var _useUnmountedRef = _interopRequireDefault(require("./useUnmountedRef"));
var _useUpdate = _interopRequireDefault(require("./useUpdate"));
var _useUpdateEffect = _interopRequireDefault(require("./useUpdateEffect"));
var _useUpdateLayoutEffect = _interopRequireDefault(require("./useUpdateLayoutEffect"));
var _useVirtualList = _interopRequireDefault(require("./useVirtualList"));
var _useWebSocket = _interopRequireDefault(require("./useWebSocket"));
var _useWhyDidYouUpdate = _interopRequireDefault(require("./useWhyDidYouUpdate"));
var _useMutationObserver = _interopRequireDefault(require("./useMutationObserver"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
