// @ant-design/pro-list/lib/style/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.techUiListActive = void 0;
exports.useStyle = useStyle;
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _cssinjs = require("@ant-design/cssinjs");
var _proProvider = require("@ant-design/pro-provider");
var techUiListActive = new _cssinjs.Keyframes('techUiListActive', {
  '0%': {
    backgroundColor: 'unset'
  },
  '30%': {
    background: '#fefbe6'
  },
  '100%': {
    backgroundColor: 'unset'
  }
});
exports.techUiListActive = techUiListActive;
var genProListStyle = function genProListStyle(token) {
  var _hover, _noSplit, _$concat5, _$concat6, _row, _token$componentCls;
  return (0, _defineProperty2.default)({}, token.componentCls, (_token$componentCls = {
    backgroundColor: 'transparent'
  }, (0, _defineProperty2.default)(_token$componentCls, "".concat(token.proComponentsCls, "-table-alert"), {
    marginBlockEnd: '16px'
  }), (0, _defineProperty2.default)(_token$componentCls, '&-row', (_row = {
    borderBlockEnd: "1px solid ".concat(token.colorSplit)
  }, (0, _defineProperty2.default)(_row, "".concat(token.antCls, "-list-item-meta-title"), {
    borderBlockEnd: 'none',
    margin: 0
  }), (0, _defineProperty2.default)(_row, '&:last-child', (0, _defineProperty2.default)({
    borderBlockEnd: 'none'
  }, "".concat(token.antCls, "-list-item"), {
    borderBlockEnd: 'none'
  })), (0, _defineProperty2.default)(_row, '&:hover', (_hover = {
    backgroundColor: 'rgba(0, 0, 0, 0.02)',
    transition: 'background-color 0.3s'
  }, (0, _defineProperty2.default)(_hover, "".concat(token.antCls, "-list-item-action"), {
    display: 'block'
  }), (0, _defineProperty2.default)(_hover, "".concat(token.antCls, "-list-item-extra"), {
    display: 'flex'
  }), (0, _defineProperty2.default)(_hover, "".concat(token.componentCls, "-row-extra"), {
    display: 'block'
  }), (0, _defineProperty2.default)(_hover, "".concat(token.componentCls, "-row-subheader-actions"), {
    display: 'block'
  }), _hover)), (0, _defineProperty2.default)(_row, '&-card', (0, _defineProperty2.default)({
    marginBlock: 8,
    marginInline: 0,
    paddingBlock: 0,
    paddingInline: 8,
    '&:hover': {
      backgroundColor: 'transparent'
    }
  }, "".concat(token.antCls, "-list-item-meta-title"), {
    flexShrink: 9,
    marginBlock: 0,
    marginInline: 0,
    lineHeight: '22px'
  })), (0, _defineProperty2.default)(_row, "&".concat(token.componentCls, "-row-editable"), (0, _defineProperty2.default)({}, "".concat(token.componentCls, "-list-item"), {
    '&-meta': {
      '&-avatar,&-description,&-title': {
        paddingBlock: 6,
        paddingInline: 0
      }
    },
    '&-action': {
      display: 'block'
    }
  })), (0, _defineProperty2.default)(_row, "&".concat(token.componentCls, "-row-selected"), {
    backgroundColor: token.colorPrimaryBgHover,
    '&:hover': {
      backgroundColor: token.colorPrimaryBgHover
    }
  }), (0, _defineProperty2.default)(_row, "&".concat(token.componentCls, "-row-type-new"), {
    animationName: techUiListActive,
    animationDuration: '3s'
  }), (0, _defineProperty2.default)(_row, "&".concat(token.componentCls, "-row-type-inline"), (0, _defineProperty2.default)({}, "".concat(token.componentCls, "-row-title"), {
    fontWeight: 'normal'
  })), (0, _defineProperty2.default)(_row, "&".concat(token.componentCls, "-row-type-top"), {
    backgroundImage: "url('https://gw.alipayobjects.com/zos/antfincdn/DehQfMbOJb/icon.svg')",
    backgroundRepeat: 'no-repeat',
    backgroundPosition: 'left top',
    backgroundSize: '12px 12px'
  }), (0, _defineProperty2.default)(_row, '&-show-action-hover', (0, _defineProperty2.default)({}, "".concat(token.antCls, "-list-item-action,\n            ").concat(token.proComponentsCls, "-card-extra,\n            ").concat(token.proComponentsCls, "-card-actions"), {
    display: 'flex'
  })), (0, _defineProperty2.default)(_row, '&-show-extra-hover', (0, _defineProperty2.default)({}, "".concat(token.antCls, "-list-item-extra"), {
    display: 'none'
  })), (0, _defineProperty2.default)(_row, '&-extra', {
    display: 'none'
  }), (0, _defineProperty2.default)(_row, '&-subheader', {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'space-between',
    height: '44px',
    paddingInline: 24,
    paddingBlock: 0,
    color: token.colorTextSecondary,
    lineHeight: '44px',
    background: 'rgba(0, 0, 0, 0.02)',
    '&-actions': {
      display: 'none'
    },
    '&-actions *': {
      marginInlineEnd: 8,
      '&:last-child': {
        marginInlineEnd: 0
      }
    }
  }), (0, _defineProperty2.default)(_row, '&-expand-icon', {
    marginInlineEnd: 8,
    display: 'flex',
    fontSize: 12,
    cursor: 'pointer',
    height: '24px',
    marginRight: 4,
    color: token.colorTextSecondary,
    '> .anticon > svg': {
      transition: '0.3s'
    }
  }), (0, _defineProperty2.default)(_row, '&-expanded', {
    ' > .anticon > svg': {
      transform: 'rotate(90deg)'
    }
  }), (0, _defineProperty2.default)(_row, '&-title', {
    marginInlineEnd: '16px',
    wordBreak: 'break-all',
    cursor: 'pointer',
    '&:hover': {
      color: token.colorPrimary
    }
  }), (0, _defineProperty2.default)(_row, '&-content', {
    position: 'relative',
    display: 'flex',
    flex: '1',
    flexDirection: 'column',
    marginBlock: 0,
    marginInline: 32
  }), (0, _defineProperty2.default)(_row, '&-subTitle', {
    color: 'rgba(0, 0, 0, 0.45)'
  }), (0, _defineProperty2.default)(_row, '&-description', {
    marginBlockStart: '4px',
    wordBreak: 'break-all'
  }), (0, _defineProperty2.default)(_row, '&-avatar', {
    display: 'flex'
  }), (0, _defineProperty2.default)(_row, '&-header', {
    display: 'flex',
    flex: '1',
    justifyContent: 'flex-start',
    h4: {
      margin: 0,
      padding: 0
    }
  }), (0, _defineProperty2.default)(_row, '&-header-title', {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'flex-start'
  }), (0, _defineProperty2.default)(_row, '&-header-option', {
    display: 'flex'
  }), (0, _defineProperty2.default)(_row, '&-checkbox', {
    width: '16px',
    marginInlineEnd: '12px'
  }), (0, _defineProperty2.default)(_row, '&-no-split', (_noSplit = {}, (0, _defineProperty2.default)(_noSplit, "".concat(token.componentCls, "-row"), {
    borderBlockEnd: 'none'
  }), (0, _defineProperty2.default)(_noSplit, "".concat(token.antCls, "-list ").concat(token.antCls, "-list-item"), {
    borderBlockEnd: 'none'
  }), _noSplit)), (0, _defineProperty2.default)(_row, '&-bordered', (0, _defineProperty2.default)({}, "".concat(token.componentCls, "-toolbar"), {
    borderBlockEnd: "1px solid ".concat(token.colorSplit)
  })), (0, _defineProperty2.default)(_row, "".concat(token.antCls, "-list-vertical"), (_$concat5 = {}, (0, _defineProperty2.default)(_$concat5, "".concat(token.componentCls, "-row"), {
    borderBlockEnd: '12px 18px 12px 24px'
  }), (0, _defineProperty2.default)(_$concat5, '&-header-title', {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'flex-start',
    justifyContent: 'center'
  }), (0, _defineProperty2.default)(_$concat5, '&-content', {
    marginBlock: 0,
    marginInline: 0
  }), (0, _defineProperty2.default)(_$concat5, '&-subTitle', {
    marginBlockStart: 8
  }), (0, _defineProperty2.default)(_$concat5, "".concat(token.antCls, "-list-item-extra"), (0, _defineProperty2.default)({
    display: 'flex',
    alignItems: 'center',
    marginInlineStart: '32px'
  }, "".concat(token.componentCls, "-row-description"), {
    marginBlockStart: 16
  })), (0, _defineProperty2.default)(_$concat5, "".concat(token.antCls, "-list-bordered ").concat(token.antCls, "-list-item"), {
    paddingInline: 0
  }), (0, _defineProperty2.default)(_$concat5, "".concat(token.componentCls, "-row-show-extra-hover"), (0, _defineProperty2.default)({}, "".concat(token.antCls, "-list-item-extra "), {
    display: 'none'
  })), _$concat5)), (0, _defineProperty2.default)(_row, "".concat(token.antCls, "-list-pagination"), {
    marginBlockStart: token.margin,
    marginBlockEnd: token.margin
  }), (0, _defineProperty2.default)(_row, "".concat(token.antCls, "-list-list"), {
    '&-item': {
      cursor: 'pointer',
      paddingBlock: 12,
      paddingInline: 12
    }
  }), (0, _defineProperty2.default)(_row, "".concat(token.antCls, "-list-vertical ").concat(token.proComponentsCls, "-list-row"), (0, _defineProperty2.default)({
    '&-header': {
      paddingBlock: 0,
      paddingInline: 0,
      borderBlockEnd: 'none'
    }
  }, "".concat(token.antCls, "-list-item"), (_$concat6 = {
    width: '100%',
    paddingBlock: 12,
    paddingInlineStart: 24,
    paddingInlineEnd: 18
  }, (0, _defineProperty2.default)(_$concat6, "".concat(token.antCls, "-list-item-meta-avatar"), {
    display: 'flex',
    alignItems: 'center',
    marginInlineEnd: 8
  }), (0, _defineProperty2.default)(_$concat6, "".concat(token.antCls, "-list-item-action-split"), {
    display: 'none'
  }), (0, _defineProperty2.default)(_$concat6, "".concat(token.antCls, "-list-item-meta-title"), {
    marginBlock: 0,
    marginInline: 0
  }), _$concat6))), _row)), _token$componentCls));
};
function useStyle(prefixCls) {
  return (0, _proProvider.useStyle)('ProList', function (token) {
    var proListToken = (0, _objectSpread2.default)((0, _objectSpread2.default)({}, token), {}, {
      componentCls: ".".concat(prefixCls)
    });
    return [genProListStyle(proListToken)];
  });
}
