// @umijs/ui/dist/p__imports__components__ViewList.async.js
(self.webpackChunk_umijs_ui=self.webpackChunk_umijs_ui||[]).push([[329],{11633:function(t,o,e){"use strict";e.r(o),e.d(o,{ViewList:function(){return W}});var l=e(65645),s=e.n(l),i=e(45990),u=e.n(i),p=e(28056),_=e.n(p),D=e(30622),C=e(65948),g=e(42890),I=e(25239),d=e(18855),c=e(26568),n=e(29088),b,E,M,T=c.styled.div(b||(b=_()([`
  .switch-container {
    margin-bottom: 1rem;
    display: flex;
    align-items: center;

    .ant-switch {
      margin-right: 0.5rem;
      background: var(--text-color);

      &-checked {
        background: var(--highlight-color);
      }
    }
  }
`]))),L=c.styled.div(E||(E=_()([`
  background: var(--bg-color);

  .ant-input-affix-wrapper {
    background: var(--bg-color);
    border-color: var(--text-color);
  }

  .ant-input {
    background: var(--bg-color);
    color: var(--text-color);
  }

  .search-icon {
    margin-right: 0.5rem;
    color: var(--text-color);
  }
`]))),U=c.styled.div(M||(M=_()([`
  margin-top: 2rem;

  .ant-list-item {
    color: var(--text-color);
    border-color: var(--text-color);
    padding-left: 1rem;
  }
`]))),O=47,S=[/node_modules/],K=function(v){return S.some(function(m){return m.test(v)})},W=function(v){var m=v.metaFile,R=(0,d.useState)(!1),y=u()(R,2),P=y[0],H=y[1],Z=(0,d.useState)(""),j=u()(Z,2),x=j[0],w=j[1],f=(0,d.useMemo)(function(){var a=m.inputs,r=Object.keys(a).reduce(function(A,V){var F=a[V].imports||[],z=F.map(function(h){return h.path}).filter(function(h){var G=h.indexOf(x)!==-1;return G&&(P||!K(h))});return[].concat(s()(A),s()(z))},[]);return Array.from(new Set(r)).sort()},[m,P,x]),N=(0,d.useMemo)(function(){var a=window.innerHeight-216,r=O*f.length;return r>a?a:r},[f]);return(0,n.jsxs)("div",{children:[(0,n.jsx)(T,{children:(0,n.jsxs)("div",{className:"switch-container",children:[(0,n.jsx)(D.Z,{onChange:H}),"show node_modules paths"]})}),(0,n.jsx)(L,{children:(0,n.jsx)(C.Z,{size:"large",value:x,onChange:function(r){return w(r.target.value)},prefix:(0,n.jsx)(c.Icon,{width:"24",height:"24",icon:"ant-design:search-outlined",className:"search-icon"})})}),(0,n.jsx)(U,{children:(0,n.jsx)(g.ZP,{bordered:!0,children:(0,n.jsx)(I.Z,{data:f,height:N,itemHeight:O,itemKey:function(r){return r},children:function(r,A){return(0,n.jsx)(g.ZP.Item,{children:(0,n.jsx)("div",{children:r})})}})})})]})}},19979:function(t,o,e){var l=e(89415);function s(i){if(Array.isArray(i))return l(i)}t.exports=s,t.exports.__esModule=!0,t.exports.default=t.exports},10424:function(t){function o(e){if(typeof Symbol!="undefined"&&e[Symbol.iterator]!=null||e["@@iterator"]!=null)return Array.from(e)}t.exports=o,t.exports.__esModule=!0,t.exports.default=t.exports},19621:function(t){function o(){throw new TypeError(`Invalid attempt to spread non-iterable instance.
In order to be iterable, non-array objects must have a [Symbol.iterator]() method.`)}t.exports=o,t.exports.__esModule=!0,t.exports.default=t.exports},65645:function(t,o,e){var l=e(19979),s=e(10424),i=e(70891),u=e(19621);function p(_){return l(_)||s(_)||i(_)||u()}t.exports=p,t.exports.__esModule=!0,t.exports.default=t.exports}}]);
