// @antv/g2plot/lib/adaptor/geometries/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.violin = exports.schema = exports.polygon = exports.point = exports.line = exports.interval = exports.edge = exports.area = void 0;
var area_1 = require("./area");
Object.defineProperty(exports, "area", { enumerable: true, get: function () { return area_1.area; } });
var edge_1 = require("./edge");
Object.defineProperty(exports, "edge", { enumerable: true, get: function () { return edge_1.edge; } });
var interval_1 = require("./interval");
Object.defineProperty(exports, "interval", { enumerable: true, get: function () { return interval_1.interval; } });
var line_1 = require("./line");
Object.defineProperty(exports, "line", { enumerable: true, get: function () { return line_1.line; } });
var point_1 = require("./point");
Object.defineProperty(exports, "point", { enumerable: true, get: function () { return point_1.point; } });
var polygon_1 = require("./polygon");
Object.defineProperty(exports, "polygon", { enumerable: true, get: function () { return polygon_1.polygon; } });
var schema_1 = require("./schema");
Object.defineProperty(exports, "schema", { enumerable: true, get: function () { return schema_1.schema; } });
var violin_1 = require("./violin");
Object.defineProperty(exports, "violin", { enumerable: true, get: function () { return violin_1.violin; } });
