// @ant-design/pro-utils/lib/components/FieldLabel/style.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useStyle = useStyle;
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _proProvider = require("@ant-design/pro-provider");
var genProStyle = function genProStyle(token) {
  var _$concat$concat, _small;
  return (0, _defineProperty2.default)({}, token.componentCls, {
    display: 'inline-flex',
    gap: '4px',
    alignItems: 'center',
    height: '30px',
    paddingBlock: 0,
    paddingInline: 8,
    fontSize: token.fontSize,
    lineHeight: '30px',
    borderRadius: '2px',
    cursor: 'pointer',
    '&:hover': {
      backgroundColor: token.colorBgTextHover
    },
    '&-active': (0, _defineProperty2.default)({
      paddingBlock: 0,
      paddingInline: 12,
      backgroundColor: token.colorBgTextHover
    }, "&".concat(token.componentCls, "-allow-clear:hover:not(").concat(token.componentCls, "-disabled)"), (_$concat$concat = {}, (0, _defineProperty2.default)(_$concat$concat, "".concat(token.componentCls, "-arrow"), {
      display: 'none'
    }), (0, _defineProperty2.default)(_$concat$concat, "".concat(token.componentCls, "-close"), {
      display: 'inline-flex'
    }), _$concat$concat)),
    '&-icon': (0, _defineProperty2.default)({
      height: '12px',
      paddingBlock: 1,
      paddingInline: 1,
      color: token.colorIcon,
      fontSize: '12px',
      verticalAlign: 'middle'
    }, "&".concat(token.componentCls, "-close"), {
      display: 'none',
      height: 14,
      alignItems: 'center',
      justifyContent: 'center',
      width: 14,
      color: token.colorBgContainer,
      fontSize: 8,
      backgroundColor: token.colorTextPlaceholder,
      borderRadius: '50%',
      '&:hover': {
        color: token.colorIcon
      }
    }),
    '&-disabled': (0, _defineProperty2.default)({
      color: token.colorTextPlaceholder,
      cursor: 'not-allowed'
    }, "".concat(token.componentCls, "-icon"), {
      color: token.colorTextPlaceholder
    }),
    '&-small': (_small = {
      height: '24px',
      paddingBlock: 0,
      paddingInline: 4,
      fontSize: token.fontSizeSM,
      lineHeight: '24px'
    }, (0, _defineProperty2.default)(_small, "&".concat(token.componentCls, "-active"), {
      paddingBlock: 0,
      paddingInline: 8
    }), (0, _defineProperty2.default)(_small, "".concat(token.componentCls, "-icon"), {
      paddingBlock: 0,
      paddingInline: 0
    }), (0, _defineProperty2.default)(_small, "".concat(token.componentCls, "-close"), {
      marginBlockStart: '-2px',
      paddingBlock: 4,
      paddingInline: 4,
      fontSize: '6px'
    }), _small),
    '&-bordered': {
      height: '32px',
      paddingBlock: 0,
      paddingInline: 12,
      border: "".concat(token.lineWidth, "px solid ").concat(token.colorBorder),
      borderRadius: '@border-radius-base'
    },
    '&-bordered&-small': {
      height: '24px',
      paddingBlock: 0,
      paddingInline: 8
    },
    '&-bordered&-active': {
      backgroundColor: token.colorBgContainer
    }
  });
};
function useStyle(prefixCls) {
  return (0, _proProvider.useStyle)('FieldLabel', function (token) {
    var proToken = (0, _objectSpread2.default)((0, _objectSpread2.default)({}, token), {}, {
      componentCls: ".".concat(prefixCls)
    });
    return [genProStyle(proToken)];
  });
}
