// @ant-design/pro-field/lib/components/TextArea/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _proProvider = require("@ant-design/pro-provider");
var _antd = require("antd");
var _react = _interopRequireDefault(require("react"));
require("antd/lib/input/style");
var _jsxRuntime = require("react/jsx-runtime");
// 兼容代码-----------

//------------
/**
 * 最基本的组件，就是个普通的 Input.TextArea
 *
 * @param
 */
var FieldTextArea = function FieldTextArea(_ref, ref) {
  var text = _ref.text,
    mode = _ref.mode,
    render = _ref.render,
    renderFormItem = _ref.renderFormItem,
    fieldProps = _ref.fieldProps;
  var intl = (0, _proProvider.useIntl)();
  if (mode === 'read') {
    var dom = /*#__PURE__*/(0, _jsxRuntime.jsx)("span", {
      ref: ref,
      style: {
        display: 'inline-block',
        padding: '4px 11px',
        lineHeight: '1.5715',
        maxWidth: '100%',
        whiteSpace: 'pre-wrap'
      },
      children: text !== null && text !== void 0 ? text : '-'
    });
    if (render) {
      return render(text, (0, _objectSpread2.default)({
        mode: mode
      }, fieldProps), dom);
    }
    return dom;
  }
  if (mode === 'edit' || mode === 'update') {
    var _dom = /*#__PURE__*/(0, _jsxRuntime.jsx)(_antd.Input.TextArea, (0, _objectSpread2.default)({
      ref: ref,
      rows: 3,
      onKeyPress: function onKeyPress(e) {
        if (e.key === 'Enter') e.stopPropagation();
      },
      placeholder: intl.getMessage('tableForm.inputPlaceholder', '请输入')
    }, fieldProps));
    if (renderFormItem) {
      return renderFormItem(text, (0, _objectSpread2.default)({
        mode: mode
      }, fieldProps), _dom);
    }
    return _dom;
  }
  return null;
};
var _default = /*#__PURE__*/_react.default.forwardRef(FieldTextArea);
exports.default = _default;
