// @ant-design/pro-card/lib/components/CheckCard/style.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.cardLoading = void 0;
exports.useStyle = useStyle;
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _cssinjs = require("@ant-design/cssinjs");
var _proProvider = require("@ant-design/pro-provider");
var proCheckCardActive = function proCheckCardActive(token) {
  return {
    backgroundColor: token.colorPrimaryBgHover,
    borderColor: token.colorPrimary
  };
};
var proCheckCardDisabled = function proCheckCardDisabled(token) {
  return (0, _defineProperty2.default)({
    backgroundColor: token.colorBgContainerDisabled,
    borderColor: token.colorBorder,
    cursor: 'not-allowed'
  }, token.componentCls, {
    '&-description': {
      color: token.colorTextDisabled
    },
    '&-title': {
      color: token.colorTextDisabled
    },
    '&-avatar': {
      opacity: '0.25'
    }
  });
};
var cardLoading = new _cssinjs.Keyframes('card-loading', {
  '0%': {
    backgroundPosition: '0 50%'
  },
  '50%': {
    backgroundPosition: '100% 50%'
  },
  '100%': {
    backgroundPosition: '0 50%'
  }
});
exports.cardLoading = cardLoading;
var genProStyle = function genProStyle(token) {
  var _token$componentCls;
  return (0, _defineProperty2.default)({}, token.componentCls, (_token$componentCls = {
    position: 'relative',
    display: 'inline-block',
    width: '320px',
    marginInlineEnd: '16px',
    marginBlockEnd: '16px',
    color: token.colorText,
    fontSize: token.fontSize,
    lineHeight: token.lineHeight,
    verticalAlign: 'top',
    backgroundColor: token.colorBgContainer,
    borderRadius: token.borderRadius,
    cursor: 'pointer',
    transition: "all 0.3s",
    '&:last-child': {
      marginInlineEnd: 0
    },
    '& + &': {
      marginInlineStart: '0 !important'
    },
    '&-bordered': {
      border: "".concat(token.lineWidth, "px solid ").concat(token.colorBorder)
    },
    '&-group': {
      display: 'inline-block'
    }
  }, (0, _defineProperty2.default)(_token$componentCls, "".concat(token.componentCls, "-loading"), {
    overflow: 'hidden',
    userSelect: 'none',
    '&-content': (0, _defineProperty2.default)({
      paddingInline: token.padding,
      paddingBlock: token.paddingSM,
      p: {
        marginBlock: 0,
        marginInline: 0
      }
    }, "".concat(token.componentCls, "-loading-block"), {
      height: '14px',
      marginBlock: '4px',
      background: "linear-gradient(90deg, rgba(54, 61, 64, 0.2), rgba(54, 61, 64, 0.4), rgba(54, 61, 64, 0.2))",
      animationName: cardLoading,
      animationDuration: '1.4s',
      animationTimingFunction: 'ease',
      animationIterationCount: 'infinite'
    })
  }), (0, _defineProperty2.default)(_token$componentCls, '&:focus', proCheckCardActive(token)), (0, _defineProperty2.default)(_token$componentCls, '&-checked', (0, _objectSpread2.default)((0, _objectSpread2.default)({}, proCheckCardActive(token)), {}, {
    '&:after': {
      position: 'absolute',
      insetBlockStart: 2,
      insetInlineEnd: 2,
      width: 0,
      height: 0,
      border: "6px solid ".concat(token.colorPrimary),
      borderBlockEnd: '6px solid transparent',
      borderInlineStart: '6px solid transparent',
      borderStartEndRadius: '2px',
      content: "''"
    }
  })), (0, _defineProperty2.default)(_token$componentCls, '&-disabled', proCheckCardDisabled(token)), (0, _defineProperty2.default)(_token$componentCls, '&[disabled]', proCheckCardDisabled(token)), (0, _defineProperty2.default)(_token$componentCls, '&-checked&-disabled', {
    '&:after': {
      position: 'absolute',
      insetBlockStart: 2,
      insetInlineEnd: 2,
      width: 0,
      height: 0,
      border: "6px solid ".concat(token.colorTextDisabled),
      borderBlockEnd: '6px solid transparent',
      borderInlineStart: '6px solid transparent',
      borderStartEndRadius: '2px',
      content: "''"
    }
  }), (0, _defineProperty2.default)(_token$componentCls, '&-lg', {
    width: 440
  }), (0, _defineProperty2.default)(_token$componentCls, '&-sm', {
    width: 212
  }), (0, _defineProperty2.default)(_token$componentCls, '&-cover', {
    paddingInline: token.paddingXXS,
    paddingBlock: token.paddingXXS,
    img: {
      width: '100%',
      height: '100%',
      overflow: 'hidden',
      borderRadius: token.borderRadius
    }
  }), (0, _defineProperty2.default)(_token$componentCls, '&-content', {
    display: 'flex',
    paddingInline: token.paddingSM,
    paddingBlock: token.padding
  }), (0, _defineProperty2.default)(_token$componentCls, '&-avatar-header', {
    display: 'flex',
    alignItems: 'center'
  }), (0, _defineProperty2.default)(_token$componentCls, '&-avatar', {
    paddingInlineEnd: 8
  }), (0, _defineProperty2.default)(_token$componentCls, '&-detail', {
    overflow: 'hidden',
    width: '100%',
    '> div:not(:last-child)': {
      marginBlockEnd: 4
    }
  }), (0, _defineProperty2.default)(_token$componentCls, '&-header', {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'space-between'
  }), (0, _defineProperty2.default)(_token$componentCls, '&-title', {
    overflow: 'hidden',
    color: token.colorTextHeading,
    fontWeight: '500',
    fontSize: token.fontSize,
    whiteSpace: 'nowrap',
    textOverflow: 'ellipsis'
  }), (0, _defineProperty2.default)(_token$componentCls, '&-description', {
    color: token.colorTextSecondary
  }), (0, _defineProperty2.default)(_token$componentCls, "&:not(".concat(token.componentCls, "-disabled)"), {
    '&:hover': {
      borderColor: token.colorPrimary
    }
  }), _token$componentCls));
};
function useStyle(prefixCls) {
  return (0, _proProvider.useStyle)('CheckCard', function (token) {
    var proListToken = (0, _objectSpread2.default)((0, _objectSpread2.default)({}, token), {}, {
      componentCls: ".".concat(prefixCls)
    });
    return [genProStyle(proListToken)];
  });
}
